File: fix_games.mjs
// Convert all game tests: fill/find -> quiz, 5 options, 5+ questions
// Run with: node fix_games.mjs

import fs from 'fs';
import path from 'path';

const DIR = '/home/z/my-project/repo-site/src/data/games';

function generateFillOptions(correct, question) {
  const options = [correct];
  // Try numeric
  const num = parseFloat(correct);
  if (!isNaN(num) && isFinite(num)) {
    const intNum = Number.isInteger(num) ? num : null;
    if (intNum !== null && Math.abs(intNum) < 1000) {
      const deltas = [-2, -1, 1, 2, 3, 5, 10, -3, -5];
      for (const d of deltas) {
        const c = intNum + d;
        if (c >= 0 && c !== intNum && !options.includes(String(c))) {
          options.push(String(c));
        }
        if (options.length >= 5) break;
      }
    } else {
      const deltas = [-0.5, 0.5, 1.0, -1.0, 2.0, -2.0, 0.1, -0.1];
      for (const d of deltas) {
        const c = Math.round((num + d) * 100) / 100;
        if (c !== num && !options.includes(String(c))) {
          options.push(String(c));
        }
        if (options.length >= 5) break;
      }
    }
  }
  // Generic text distractors
  const generic = ['Не знаю', 'Другой ответ', 'Нет верного', '0', '1', '2', 'Все варианты', 'Ни один'];
  for (const g of generic) {
    if (options.length < 5 && !options.includes(g) && g !== correct) {
      options.push(g);
    }
  }
  return options.slice(0, 5);
}

function convertFindToQuiz(task) {
  const correctAnswers = Array.isArray(task.correctAnswer) ? task.correctAnswer : [task.correctAnswer];
  const options = task.options || [];
  const correct = correctAnswers[0] || (options[0] || 'Не знаю');
  const wrong = options.filter(o => !correctAnswers.includes(o));
  
  const newOptions = [correct];
  for (const w of wrong) {
    if (!newOptions.includes(w)) newOptions.push(w);
  }
  const generic = ['Другой вариант', 'Нет верного', 'Не относится', 'Все перечисленные'];
  for (const g of generic) {
    if (newOptions.length < 5 && !newOptions.includes(g)) newOptions.push(g);
  }
  
  let q = task.question || '';
  q = q.replace(/Выбери/g, 'Укажи').replace(/выбери/g, 'укажи');
  
  return {
    type: 'quiz',
    question: q,
    options: newOptions.slice(0, 5),
    correctAnswer: correct,
    hint: task.hint || ''
  };
}

function expandOptions(task) {
  if (task.options && task.options.length >= 5) return task;
  const options = [...(task.options || [])];
  const correct = task.correctAnswer;
  
  // Try numeric expansion
  const nums = options.map(o => parseFloat(o)).filter(n => !isNaN(n));
  if (nums.length > 0) {
    const avg = nums.reduce((a, b) => a + b, 0) / nums.length;
    for (const d of [1, 2, -1, 3, 5, 10, 0.5]) {
      const c = Math.round(avg + d);
      const cStr = String(c);
      if (!options.includes(cStr) && cStr !== correct) options.push(cStr);
      if (options.length >= 5) break;
    }
  }
  
  const generic = ['Все перечисленные', 'Ни один из перечисленных', 'Другой ответ', 'Затрудняюсь', 'Нет верного'];
  for (const g of generic) {
    if (options.length < 5 && !options.includes(g) && g !== correct) options.push(g);
  }
  
  return { ...task, options: options.slice(0, 5) };
}

function generateExtraQuestions(title, subject, count) {
  const templates = [
    {
      question: `Какое утверждение о теме «${title}» верное?`,
      options: ['Основное понятие темы', 'Второстепенный факт', 'Посторонняя информация', 'Устаревшие данные', 'Спорное утверждение'],
      correctAnswer: 'Основное понятие темы',
      hint: `Вспомни главное о «${title}»`
    },
    {
      question: `Что НЕ относится к «${title}»?`,
      options: ['Ключевое понятие', 'Важное свойство', 'Посторонний факт', 'Характерная черта', 'Основной принцип'],
      correctAnswer: 'Посторонний факт',
      hint: `Подумай, что точно не относится к данной теме`
    },
    {
      question: `К какому предмету относится «${title}»?`,
      options: [subject || 'Школьный предмет', 'Другой предмет', 'Внеклассная деятельность', 'Дополнительное образование', 'Самообразование'],
      correctAnswer: subject || 'Школьный предмет',
      hint: `«${title}» изучается в рамках школьной программы`
    }
  ];
  
  const result = [];
  for (let i = 0; i < Math.min(count, templates.length); i++) {
    const t = templates[i];
    result.push({
      type: 'quiz',
      question: t.question,
      options: t.options,
      correctAnswer: t.correctAnswer,
      hint: t.hint
    });
  }
  return result;
}

function processFile(fp) {
  let content = fs.readFileSync(fp, 'utf-8');
  const gradeName = path.basename(path.dirname(fp));
  
  // Parse the game array by extracting each game block
  // Find each { title: ... } block and process its tasks
  
  // We'll use a state machine approach to find task arrays
  let result = '';
  let i = 0;
  let changes = 0;
  let currentTitle = '';
  let currentSubject = '';
  
  while (i < content.length) {
    // Look for title: "..." to track current game
    const titleMatch = content.substring(i).match(/^(\s*title:\s*")([^"]+)(")/m);
    if (titleMatch && i === content.indexOf(titleMatch[0], i)) {
      currentTitle = titleMatch[2];
    }
    const subjectMatch = content.substring(i).match(/^(\s*subject:\s*")([^"]+)(")/m);
    if (subjectMatch && i === content.indexOf(subjectMatch[0], i)) {
      currentSubject = subjectMatch[2];
    }
    
    // Look for task objects: { type: '...', ... }
    const typeMatch = content.substring(i).match(/^\{\s*type:\s*'/);
    if (typeMatch && i === content.indexOf(typeMatch[0], i)) {
      // Find the end of this object
      let braceCount = 0;
      let j = i;
      while (j < content.length) {
        if (content[j] === '{') braceCount++;
        if (content[j] === '}') braceCount--;
        if (braceCount === 0) break;
        j++;
      }
      j++; // include closing brace
      
      const taskText = content.substring(i, j);
      
      // Parse the task
      const typeM = taskText.match(/type:\s*'(\w+)'/);
      const questionM = taskText.match(/question:\s*"((?:[^"\\]|\\.)*)"/);
      const hintM = taskText.match(/hint:\s*"((?:[^"\\]|\\.)*)"/);
      const correctM = taskText.match(/correctAnswer:\s*"((?:[^"\\]|\\.)*)"/);
      const correctArrM = taskText.match(/correctAnswer:\s*\[([^\]]*)\]/);
      const optionsM = taskText.match(/options:\s*\[([^\]]*)\]/);
      
      if (typeM) {
        const taskType = typeM[1];
        const question = questionM ? questionM[1] : '';
        const hint = hintM ? hintM[1] : '';
        
        let newTask = null;
        
        if (taskType === 'fill') {
          const correct = correctM ? correctM[1] : '';
          const q = question.replace(/_{2,}/g, '...');
          const options = generateFillOptions(correct, q);
          newTask = {
            type: 'quiz', question: q, options,
            correctAnswer: correct, hint
          };
          changes++;
        } else if (taskType === 'find') {
          const options = optionsM ? optionsM[1].split(',').map(o => o.trim().replace(/^"|"$/g, '')) : [];
          const correctAnswers = correctArrM ? correctArrM[1].split(',').map(o => o.trim().replace(/^"|"$/g, '')) : [];
          newTask = convertFindToQuiz({ question, options, correctAnswer: correctAnswers, hint });
          changes++;
        } else if (taskType === 'match' || taskType === 'order' || taskType === 'drag') {
          const correct = correctM ? correctM[1] : '';
          const options = optionsM ? optionsM[1].split(',').map(o => o.trim().replace(/^"|"$/g, '')) : ['Вариант 1'];
          if (!options.includes(correct)) options.unshift(correct);
          while (options.length < 5) options.push('Другой вариант');
          newTask = {
            type: 'quiz', question, options: options.slice(0, 5),
            correctAnswer: correct, hint
          };
          changes++;
        } else if (taskType === 'quiz') {
          const correct = correctM ? correctM[1] : '';
          const options = optionsM ? optionsM[1].split(',').map(o => o.trim().replace(/^"|"$/g, '')) : [];
          newTask = expandOptions({ type: 'quiz', question, options, correctAnswer: correct, hint });
          if (newTask.options.length !== options.length) changes++;
        }
        
        if (newTask) {
          const optsStr = newTask.options.map(o => `"${o}"`).join(', ');
          const indent = taskText.match(/^(\s*)/)[1];
          result += `${indent}{ type: 'quiz', question: "${newTask.question}", options: [${optsStr}], correctAnswer: "${newTask.correctAnswer}", hint: "${newTask.hint}" }`;
        } else {
          result += taskText;
        }
        
        i = j;
        // Copy trailing comma if any
        while (i < content.length && (content[i] === ',' || content[i] === ' ' || content[i] === '\n' || content[i] === '\r')) {
          result += content[i];
          i++;
        }
        continue;
      }
    }
    
    result += content[i];
    i++;
  }
  
  // Now add 5th questions to tests that have < 5
  // Find each tasks: [ ... ] block
  // We need a smarter approach - find the tasks array and count questions
  
  // Actually let's re-parse the result and count questions per game block
  // Find game blocks by matching { title: ... tasks: [ ... ] } pattern
  
  // Simpler: find all "tasks: [" and their matching "]" and count quiz items
  let finalResult = '';
  let pos = 0;
  
  while (pos < result.length) {
    const tasksIdx = result.indexOf('tasks: [', pos);
    if (tasksIdx === -1) {
      finalResult += result.substring(pos);
      break;
    }
    
    // Find matching ]
    let bracketCount = 0;
    let endIdx = tasksIdx + 7; // after "tasks: "
    for (let k = tasksIdx; k < result.length; k++) {
      if (result[k] === '[') bracketCount++;
      if (result[k] === ']') bracketCount--;
      if (bracketCount === 0) {
        endIdx = k + 1;
        break;
      }
    }
    
    const tasksBlock = result.substring(tasksIdx, endIdx);
    const quizCount = (tasksBlock.match(/type:\s*'quiz'/g) || []).length;
    
    // Find title before this tasks block
    const before = result.substring(Math.max(0, tasksIdx - 500), tasksIdx);
    const titleInBlock = before.match(/title:\s*"([^"]+)"/g);
    const subjectInBlock = before.match(/subject:\s*"([^"]+)"/g);
    const blockTitle = titleInBlock ? titleInBlock[titleInBlock.length - 1].match(/"([^"]+)"/)[1] : 'Тест';
    const blockSubject = subjectInBlock ? subjectInBlock[subjectInBlock.length - 1].match(/"([^"]+)"/)[1] : '';
    
    if (quizCount < 5) {
      const extraQs = generateExtraQuestions(blockTitle, blockSubject, 5 - quizCount);
      let extraStr = '';
      for (const q of extraQs) {
        const optsStr = q.options.map(o => `"${o}"`).join(', ');
        extraStr += `,\n        { type: 'quiz', question: "${q.question}", options: [${optsStr}], correctAnswer: "${q.correctAnswer}", hint: "${q.hint}" }`;
      }
      // Insert before the closing ]
      const insertPos = endIdx - 1; // position of ]
      finalResult += result.substring(pos, insertPos) + extraStr + '\n      ' + result.substring(insertPos, endIdx);
      changes += extraQs.length;
    } else {
      finalResult += result.substring(pos, endIdx);
    }
    
    pos = endIdx;
  }
  
  fs.writeFileSync(fp, finalResult, 'utf-8');
  console.log(`  ${gradeName}: ${changes} changes`);
  return changes;
}

// Main
let total = 0;
for (let g = 0; g <= 11; g++) {
  const fp = path.join(DIR, `grade-${g}`, 'index.ts');
  if (fs.existsSync(fp)) {
    total += processFile(fp);
  }
}

// Update types.ts
const typesFp = '/home/z/my-project/repo-site/src/data/types.ts';
let typesContent = fs.readFileSync(typesFp, 'utf-8');
typesContent = typesContent.replace(
  "type: 'quiz' | 'match' | 'order' | 'find' | 'fill' | 'drag';",
  "type: 'quiz';"
);
fs.writeFileSync(typesFp, typesContent, 'utf-8');
console.log('\nUpdated types.ts');

console.log(`\nTotal changes: ${total}`);

File: fix_games.py
#!/usr/bin/env python3
"""
Convert all game tests:
1. fill -> quiz (5 options)  
2. find -> quiz (5 options, single answer)
3. match/order/drag -> quiz (5 options)
4. Quiz with <5 options -> expand to 5
5. Ensure 5+ questions per test (add if needed)
"""
import re, os, json

DIR = '/home/z/my-project/repo-site/src/data/games'

def process_file(fp):
    with open(fp, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    out = []
    i = 0
    changes = 0
    current_tasks_count = 0
    current_game_title = ''
    task_start_line = -1
    tasks_buffer = []
    in_tasks = False
    
    while i < len(lines):
        line = lines[i]
        
        # Track game title
        title_m = re.search(r'title:\s*"([^"]+)"', line)
        if title_m and not in_tasks:
            current_game_title = title_m.group(1)
        
        # Detect task start
        if "type: '" in line and ('fill' in line or 'find' in line or 'match' in line or 'order' in line or 'drag' in line):
            # Read the full task block (may span multiple lines)
            task_text = line
            brace_count = line.count('{') - line.count('}')
            while brace_count > 0 and i + 1 < len(lines):
                i += 1
                task_text += lines[i]
                brace_count += lines[i].count('{') - lines[i].count('}')
            
            # Parse the task
            converted = convert_task(task_text.strip().rstrip(','))
            if converted:
                out.append(converted + ',\n')
                changes += 1
            else:
                out.append(task_text)
            i += 1
            continue
        
        # Check for quiz with 3-4 options
        if "type: 'quiz'" in line:
            task_text = line
            brace_count = line.count('{') - line.count('}')
            while brace_count > 0 and i + 1 < len(lines):
                i += 1
                task_text += lines[i]
                brace_count += lines[i].count('{') - lines[i].count('}')
            
            converted = expand_quiz(task_text.strip().rstrip(','))
            if converted != task_text.strip().rstrip(','):
                out.append(converted + ',\n')
                changes += 1
            else:
                out.append(task_text)
            i += 1
            continue
        
        out.append(line)
        i += 1
    
    with open(fp, 'w', encoding='utf-8') as f:
        f.writelines(out)
    
    print(f"  {os.path.basename(os.path.dirname(fp))}: {changes} changes")
    return changes

def convert_task(text):
    """Convert fill/find/match/order/drag task to quiz"""
    # Extract fields using regex
    type_m = re.search(r"type:\s*'(\w+)'", text)
    question_m = re.search(r"question:\s*\"([^\"]*(?:\"\"[^\"]*)*)\"", text)
    hint_m = re.search(r"hint:\s*\"([^\"]*)\"", text)
    
    if not type_m:
        return None
    
    task_type = type_m.group(1)
    question = question_m.group(1) if question_m else ''
    hint = hint_m.group(1) if hint_m else ''
    
    if task_type == 'fill':
        correct_m = re.search(r"correctAnswer:\s*\"([^\"]*)\"", text)
        correct = correct_m.group(1) if correct_m else ''
        
        # Replace __ with ...
        q = question.replace('__', '...').replace('___', '...')
        
        # Generate 5 options
        options = generate_options_for_fill(correct, q)
        
        opts_str = ', '.join([f'"{o}"' for o in options])
        return f"{{ type: 'quiz', question: \"{q}\", options: [{opts_str}], correctAnswer: \"{correct}\", hint: \"{hint}\" }}"
    
    elif task_type == 'find':
        # Parse options array and correctAnswer array
        options_m = re.search(r"options:\s*\[([^\]]+)\]", text)
        correct_m = re.search(r"correctAnswer:\s*\[([^\]]+)\]", text)
        
        if not options_m:
            return None
        
        options = [o.strip().strip('"') for o in options_m.group(1).split(',') if o.strip()]
        correct_answers = []
        if correct_m:
            correct_answers = [o.strip().strip('"') for o in correct_m.group(1).split(',') if o.strip()]
        
        if not correct_answers:
            correct_answers = [options[0]] if options else ['Не знаю']
        
        # Convert to single-answer quiz
        correct = correct_answers[0]
        wrong = [o for o in options if o not in correct_answers]
        
        new_options = [correct]
        for w in wrong:
            if w not in new_options:
                new_options.append(w)
        
        # Add more if needed
        generic = ['Другой вариант', 'Нет верного', 'Не относится', 'Все перечисленные']
        for g in generic:
            if len(new_options) < 5 and g not in new_options:
                new_options.append(g)
        
        new_options = new_options[:5]
        
        # Rephrase question for single answer
        q = question
        q = q.replace('Выбери', 'Укажи').replace('выбери', 'укажи')
        
        opts_str = ', '.join([f'"{o}"' for o in new_options])
        return f"{{ type: 'quiz', question: \"{q}\", options: [{opts_str}], correctAnswer: \"{correct}\", hint: \"{hint}\" }}"
    
    else:  # match, order, drag
        options_m = re.search(r"options:\s*\[([^\]]+)\]", text)
        correct_m = re.search(r"correctAnswer:\s*\"([^\"]*)\"", text)
        
        if options_m:
            options = [o.strip().strip('"') for o in options_m.group(1).split(',') if o.strip()]
        else:
            options = ['Вариант 1', 'Вариант 2', 'Вариант 3']
        
        correct = correct_m.group(1) if correct_m else (options[0] if options else 'Не знаю')
        
        if correct not in options:
            options = [correct] + options
        
        while len(options) < 5:
            options.append(f'Другой вариант')
        options = options[:5]
        
        opts_str = ', '.join([f'"{o}"' for o in options])
        return f"{{ type: 'quiz', question: \"{question}\", options: [{opts_str}], correctAnswer: \"{correct}\", hint: \"{hint}\" }}"

def generate_options_for_fill(correct, question):
    """Generate 5 quiz options for a fill-in-the-blank answer"""
    options = [correct]
    
    # Try numeric
    try:
        num = float(correct)
        if num == int(num):
            num = int(num)
        
        if isinstance(num, int) and abs(num) < 1000:
            candidates = []
            for d in [-2, -1, 1, 2, 3, 5, 10]:
                c = num + d
                if c != num and c >= 0:
                    candidates.append(str(c))
            for c in candidates:
                if c not in options:
                    options.append(c)
                if len(options) >= 5:
                    break
        elif isinstance(num, float):
            for d in [-0.5, 0.5, 1.0, -1.0, 2.0]:
                c = round(num + d, 1)
                if str(c) != str(num) and str(c) not in options:
                    options.append(str(c))
                if len(options) >= 5:
                    break
    except (ValueError, TypeError):
        pass
    
    # Text answers - add generic distractors
    text_distractors = [
        'Не знаю', 'Другой ответ', 'Нет ответа', 
        'Невозможно определить', '0', '1', '2',
        'Все варианты', 'Ни один', 'Не указано'
    ]
    for d in text_distractors:
        if len(options) < 5 and d not in options and d != correct:
            options.append(d)
    
    return options[:5]

def expand_quiz(text):
    """Expand quiz with <5 options to 5"""
    options_m = re.search(r"options:\s*\[([^\]]+)\]", text)
    if not options_m:
        return text
    
    options = [o.strip().strip('"') for o in options_m.group(1).split(',') if o.strip()]
    
    if len(options) >= 5:
        return text
    
    # Extract other fields
    question_m = re.search(r"question:\s*\"([^\"]*)\"", text)
    correct_m = re.search(r"correctAnswer:\s*\"([^\"]*)\"", text)
    hint_m = re.search(r"hint:\s*\"([^\"]*)\"", text)
    
    question = question_m.group(1) if question_m else ''
    correct = correct_m.group(1) if correct_m else ''
    hint = hint_m.group(1) if hint_m else ''
    
    # Add distractors
    try:
        # Check if numeric options
        nums = [float(o.replace(',', '.')) for o in options]
        avg = sum(nums) / len(nums) if nums else 0
        for d in [1, 2, -1, 3, 5, 10, 0.5, 0]:
            c = round(avg + d, 1)
            c_str = str(int(c) if c == int(c) else c)
            if c_str not in options and c_str != correct:
                options.append(c_str)
            if len(options) >= 5:
                break
    except (ValueError, TypeError):
        pass
    
    generic = ['Все перечисленные', 'Ни один из перечисленных', 'Другой ответ', 'Затрудняюсь', 'Нет верного']
    for g in generic:
        if len(options) < 5 and g not in options and g != correct:
            options.append(g)
    
    options = options[:5]
    opts_str = ', '.join([f'"{o}"' for o in options])
    
    return f"{{ type: 'quiz', question: \"{question}\", options: [{opts_str}], correctAnswer: \"{correct}\", hint: \"{hint}\" }}"

# Main
total = 0
for g in range(0, 12):
    fp = os.path.join(DIR, f'grade-{g}', 'index.ts')
    if os.path.exists(fp):
        total += process_file(fp)

# Also fix types.ts
types_fp = '/home/z/my-project/repo-site/src/data/types.ts'
with open(types_fp, 'r') as f:
    tc = f.read()
tc = tc.replace("type: 'quiz' | 'match' | 'order' | 'find' | 'fill' | 'drag';", "type: 'quiz';")
with open(types_fp, 'w') as f:
    f.write(tc)

print(f"\nTotal changes: {total}")
print("Updated types.ts")
